### ISSUE TITLE
always offer to pull latest image when creating a toolbox

### ISSUE THREAD

**Issue by juhp** [containers/toolbox#873]
**Is your feature request related to a problem? Please describe.**
We frequently see bug reports caused by people having an old toolbox image cached locally.

**Describe the solution you'd like**
`toolbox create` should either offer to pull or always pull down the latest image being requested .
For extra points it could check if a newer image exists first.

**Additional context**
I am imagining something like this:
```
$ toolbox create
A newer fedora-toolbox:34 image is available, do you want to download it? [Y/n]
```

I think this is the least surprising behavior (though personally I would always want to pull the latest image).
A more conservative solution could be to add a `--pull` flag.

**Comment by debarshiray** [comment]
> We frequently see bug reports caused by people having an
> old toolbox image cached locally.

Umm... do you have some examples of these bug reports?

I know that people struggle to switch from Rawhide to a freshly branched unstable Fedora at the branching point. Anything else?

Have we ever pushed out badly broken images for stable Fedoras? Ultimately users can always 'dnf update' their containers, can't they?

**Comment by juhp** [comment]
I will have to go hunting for examples: there should be enough around.

Conversely, why would you not want to create a new toolbox with the latest image?

**Comment by debarshiray** [comment]
> (Also `filesystem` can't be dnf updated.)

This one is https://github.com/containers/toolbox/issues/643, which should be fixed now.

**Comment by debarshiray** [comment]
> Conversely, why would you not want to create a new toolbox with
> the latest image?

If there were no other trade-offs, yes, I would always want to use the latest image.

However, if a user keeps accumulating a large number of images, then:
* It can negatively impact disk space.
* It can clutter up the `toolbox list` output and make it difficult to get a feel for what images are locally available.

These are just trade-offs, and we need to figure out what the right balance is. That's why I was curious about examples of bugs caused by stale images.

There could be other ways of looking at this too:
* Sharing the DNF caches to reduce network traffic.
* Prompting the user that a new image is now available, so that they can manually migrate.

**Comment by oturpe** [comment]
> > We frequently see bug reports caused by people having an
> > old toolbox image cached locally.
> 
> Umm... do you have some examples of these bug reports?
> 
> I know that people struggle to switch from Rawhide to a freshly branched unstable Fedora at the branching point. Anything else?
> 
> Have we ever pushed out badly broken images for stable Fedoras? Ultimately users can always 'dnf update' their containers, can't they?

1. At a time when `toolbox:36` image is Rawhide,
2. need a Rawhide toolbox, get it with `toolbox create --release 36`.
3. Time passes, Fedora 36 is branched and `toolbox:36` is re-defined to mean that.
4. Need a Fedora 36 toolbox, try to get it with `toolbox create --release 36`.
5. Expected result: get a Fedora 36 toolbox
6. Actual result: get a Rawhide toolbox

The image was not badly broken at the time is was created,
but the resulting container is badly broken,
because dnf does not point to repositories of the requested release.

I have reported this before [in this comment](https://github.com/containers/toolbox/issues/722#issuecomment-880454223).
Would you like to have this logged separately also?

Apart from automatically pulling, or prompting for pulling, new images,
this could also be solved through #646.

**Comment by debarshiray** [comment]
Switching from Fedora Rawhide to Branched is already a known and understood issue (and isn't specific to containers).  I was looking for other problems that we have come across over time.

**Comment by oturpe** [comment]
> Switching from Fedora Rawhide to Branched is already a known and understood issue (and isn't specific to containers). I was looking for other problems that we have come across over time.

My example is not about switching from Rawhide to Branched.
At step 2, the user needs, and gets, a new Rawhide container.
It can stay Rawhide forever without affecting the rest of the steps.
At step 4, the user needs a new Fedora 36 container.
But because Toolbx
a) first defines `--release 36` to mean Rawhide, but later re-defines it to mean Fedora 36, and
b) does not pull latest images on `create`,
the user unexpectedly gets Rawhide in step 4.

Doing what is proposed in this issue would fix b),
fixing #646 would fix a).
Fix one, and the problem disappears.

**Comment by debarshiray** [comment]
Yes, https://github.com/containers/toolbox/issues/646 is a much safer change to implement.

**Comment by debarshiray** [comment]
I found myself thinking about this today.

We might be able to solve the problem of garbage collecting older images, if we can figure out how to rebase existing containers on top of newer images (see https://github.com/containers/toolbox/issues/1358).

While playing with this today, I found some weird quirks.  I don't know if they would be considered Podman bugs or if they are expected behaviour, but they might be surprising anyway.  So, I will leave them here to save my future self and others some head-scratching.  :)

I had an existing `fedora-toolbox-42` container created from an older version of the `registry.fedoraproject.org/fedora-toolbox:42` image.  Then I did a `podman pull ...` on the same image, and got:
```ShellSession
$ toolbox list
IMAGE ID      IMAGE NAME                                     CREATED
6fddd81bb158  <none>                                         12 days ago
3aa55ef0b00a  registry.fedoraproject.org/fedora-toolbox:42   16 hours ago

CONTAINER ID  CONTAINER NAME     CREATED      STATUS  IMAGE NAME
31453ee9f926  fedora-toolbox-42  12 days ago  exited  registry.fedoraproject.org/fedora-toolbox:42
```

Notice how the older image lost its name (in `podman images`, etc.), but the container still remembers it (in `podman ps`, etc.):
```ShellSession
$ podman inspect --format '{{.Image}} {{.ImageName}}' --type container fedora-toolbox-42
6fddd81bb1585137fc3550846d5aa6b7cc99b188ff07bac37e35ae46d3495fdf registry.fedoraproject.org/fedora-toolbox:42
```

Same in `/run/.containerenv`:
```ShellSession
⬢ $ cat /run/.containerenv 
engine="podman-5.5.2"
name="fedora-toolbox-42"
id="31453ee9f92652b3065d34a10f1dbbf84a84e2e90eea3572825b88c20a041f83"
image="registry.fedoraproject.org/fedora-toolbox:42"
imageid="6fddd81bb1585137fc3550846d5aa6b7cc99b188ff07bac37e35ae46d3495fdf"
rootless=1
graphRootMounted=1
```
